test/functional/test-shell.md

```diff
@@ -0,0 +1,187 @@
+like `BitcoinTestFramework`, the TestShell allows the user to:
+
+* Manage regtest bitcoind subprocesses.
+* Access RPC interfaces of the underlying bitcoind instances.
+* Log events to the functional test logging utility.
+
+The `TestShell` can be useful in interactive environments where it is necessary
+to extend the object lifetime of the underlying `BitcoinTestFramework` between
+user inputs. Such environments include the Python3 command line interpreter or
+[Jupyter](https://jupyter.org/) notebooks running a Python3 kernel.
+
+## 1. Requirements
+
+* Python3
+* `bitcoind` built in the same repository as the TestShell.
+
+## 2. Importing TestShell from the Bitcoin Core repository
+
+We can import the TestShell by adding the path of the Bitcoin Core
+`test_framework` module to the beginning of the PATH variable, and then
+importing the `TestShell` class from the `test_shell` sub-package.
+
+```
+>>> import sys
+>>> sys.path.insert(0, "/path/to/bitcoin/test/functional")
+>>> from test_framework.test_shell import TestShell
+```
+
+The following TestShell methods manage the lifetime of the underlying bitcoind
+processes and logging utilities.
+
+* `TestShell.setup()`
+* `TestShell.shutdown()`
+
+The TestShell inherits all BitcoinTestFramework members and methods, such as:
+* `TestShell.nodes[index].rpc_method()`
+* `TestShell.log.info("Custom log message")`
+
+The following sections demonstrate how to initialize, run, and shut down a
+TestShell object.
+
+## 3. Initializing a TestShell object
+
+```
+>>> test = TestShell()
+>>> test.setup(num_nodes=2)
+20XX-XX-XXTXX:XX:XX.XXXXXXX TestFramework (INFO): Initializing test directory /path/to/bitcoin_func_test_XXXXXXX
+```
+The TestShell forwards all functional test parameters of the parent Bitcoin
+TestFramework object. The full set of argument keywords which can be used to
+initialize the TestShell can be found in [section
+#6](#custom-testshell-parameters) of this document.
+
+**Note: Running multiple instances of TestShell is not allowed.** Running a
+single process also ensures that logging remains consolidated in the same
+temporary folder. If you need more bitcoind nodes than set by default (1),
+simply increase the `num_nodes` parameter during setup.
+
+```
+>>> test2 = TestShell()
+>>> test2.setup()
+TestShell is already running!
+```
+
+## 4. Interacting with the TestShell
+
+Unlike the BitcoinTestFramework class, the TestShell keeps the underlying
+Bitcoind subprocesses (nodes) and logging utilities running until the user
+explicitly shuts down the TestShell object.
+
+During the time between the `setup` and `shutdown` calls, all `bitcoind` node
+processes and BitcoinTestFramework convenience methods can be accessed
+interactively.
+
+**Example: Mining a regtest chain**
+
+By default, the TestShell nodes are initialized with a clean chain. This means
+that each node of the TestShell is initialized with a block height of 0.
```

Comment:
I'm seeing 200 blocks by default:
```python
((HEAD detached at origin/pr/17288))$ python3
Python 3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> from test_framework.test_shell import TestShell
>>> test = TestShell()
>>> test.setup(num_nodes=2)
2019-11-04T12:07:13.890000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_d123oydj
>>> test2 = TestShell()
>>> test2.setup()
TestShell is already running!
>>> test.nodes[0].getblockchaininfo()["blocks"]
200
>>> address = test.nodes[0].getnewaddress()
>>> print(address)
bcrt1qq5pfeq0nqk8s305p29jjkn5203kjc84lgjxgf4
```
Second try without instantiating a second shell.
```python
((HEAD detached at origin/pr/17288))$ python3
Python 3.7.3 (default, Apr  3 2019, 05:39:12) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> from test_framework.test_shell import TestShell
>>> test = TestShell()
>>> test.setup(num_nodes=2)
2019-11-04T12:10:24.474000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_su_r79l1
>>> test.nodes[0].getblockchaininfo()["blocks"]
200
```

Reply:
Oops, I forgot to update that. The `setup_clean_chain` default is now `False`, since we are inheriting it from the `BitcoinTestFramework` parent class. Will update in docs, Thanks!!

Reply:
Fixed in documentation:`test.setup(num_nodes=2, setup_clean_chain=True)`.